숫자야구 [STEP2] Kiseok, Hisop

Hi-sop opened the pull request:
@leezion94  
Kiseok, Hisop 숫자야구 STEP2 PR입니다.

## 고민되었던 점
- UserInput이 옳은 입력인지 판별하는 로직 구현  
    -  처음엔 map함수를 통하여 userinput을 [Int]에 할당하려고 했으나 실패했습니다.
    -  대신 for문 안에서 옵셔널 언래핑을 하는 방식을 선택했습니다.
    - userInput을 [Int] 형태로 반환하는 함수 구현하고 이 함수를 이용하여 제시된 조건들을 충족하는지 판단하는 로직을 guard문을 활용하여 구현하였습니다.
- Error처리  
    - 에러가 발생했을 때 출력되는 값은 "입력이 잘못되었습니다"로  고정되어있어 do-try-catch와 Error를 굳이 사용해야하나 고민했습니다.
    -  사용자의 입력을 검증하는 코드를 짠 후 디버깅과정에서 적절하게 통과되는지 확인하는데 에러코드가 큰 도움이 되어 사용하기로 결정했습니다 -> 제약사항 확인 못한 부분이라 옵셔널 반환으로 변경했습니다.

- if와 guard의 선택  
    - 해당 블럭 이후에도 언래핑한 값을 사용하게 된다면 guard let을 사용했습니다.
    - 조건을 통해 해당 함수 혹은 반복문을 종료하게 되는 케이스엔 guard를 사용했습니다.

## 조언을 얻고 싶은 부분

- Error를 핸들링하는 일반적인 방법
- 에러 타입을 커스텀하지 않아도 Error를 전달할 수 있는지
- 커밋을 남기는 기준

Commit:
feat: 오류 case작성, 사용자 메뉴 출력 함수 구현, 유저 메뉴 선택 함수 구현

carti1108 committed:
feat: 사용자 입력 prepare 추가 및 기능 별 함수 분리

printMenu 함수가 문자열만 출력하도록 변경
사용자 입력에 따른 실행을 runProgram이 담당하도록 변경
사용자 입력 prepare에 옵셔널 강제추출은 나중에 수정 할 예정
오류 enum 삭제

Commit:
docs: 플로우차트 STEP2 기준으로 변경

Commit:
feat: isValidInput 함수 추가

userInput이 Int인지 판단하는 함수인 isValidInput 추가

carti1108 committed:
feat: 사용자의 입력을 검증하는 코드 추가

isValidInput함수 삭제
에러 케이스 추가
사용자가 입력한 숫자가 3개가 아니면 에러
사용자가 중복된 숫자를 입력했다면 에러
사용자가 1의자리 숫자 범위를 벗어나면 에러
사용자가 숫자를 입력하지 않았다면 에러

Commit:
style: 주석삭제 및 일부 네이밍 변경

statement를 isRunning 으로 변경
개행 등 일부 컨벤션 수정

carti1108 committed:
style: if문을 guard문으로 변경

Commit:
refactor: Error 열거형 삭제 및 입력검증함수 반환값 수정

try-do-catch를 통한 에러 핸들링을 옵셔널 반환으로 변경

LeeZion94 commented:
안녕하세요 히솝, 기석 기간내에 요구사항에 맞는 프로그램을 구현하시느라 고생하셨습니다.
우선 저번 PR에서 말씀드렸던 부분들을 이번 PR에서는 많이 보완하시고 적용하셔서 감사합니다.
작성된 코드를 살펴보면서 많은 고민들이 보였고 더 나은 코드를 작성하기 위해 여러 생각들을 하신게 나타나는 코드였습니다.

다만, Step1의 PR과 비교했을 때 전체적으로 코드를 작성하는 부분에 있어서 개인적으로는 조금 급하게 작성한게 아닌가하는 생각이 들었습니다. 물론 프로젝트의 기간이 정해져있기 때문에 같은 캠퍼로서 이해하지못하는 부분은 아닙니다만, 현재의 Step에서도 그리고 앞으로의 Step에서도 프로젝트의 진도보다는 과제를 구현하는데 있어서 필요한 요소들을 생각해보고 공식문서를 참고하고 여러 코드들을 살펴보며 고민하고 배우는 일을 우선시 하는 것이 훨씬 더 중요한 일 임을 알아주셨으면 좋겠습니다.

위와 같은 부분들만 조금 더 고려해주신다면 앞으로의 캠프생활에 있어서도 문제없이 많은 것들을 배우시며 성장해나가실 분들이라고 생각됩니다. 다시 한번 한 주 동안 고생 정말 많으셨습니다.

### 고민되었던 점
1. UserInput이 옳은 입력인지 판별하는 로직 구현
- 우선 칭찬드리고 싶은 부분은 발생할 수 잇는 대부분의 예외처리가 고려되어있는 코드라는 점입니다.
    하지만 예외처리를 다루고 있는 부분이 한 곳에 존재하는 것이 아닌 파편화 되어있는 점이 조금은 아쉽습니다. 예외처리를 하는 곳을 한 곳으로 모은다면 존재할 수 있는 예외사항에 있어서 놓치지 않고 분명히 잡아 낼 수 있고 코드의 가독성 및 유지 보수를 좀 더 용이하게 할 수 있기 때문입니다.

- 또한 언급해주신 부분 처럼 현재 error는 "입력이 잘못되었습니다" 한 개로 처리가 되고 있기 때문에 다른방법으로 error를 처리하신 것이 아닌 nil을 반환하는 것으로 error을 판단하고 있는 것으로 보입니다. 다뤄야하는 error 자체가 1개였기 때문에 잘생각해주신 부분이라고 생각합니다. 하지만 error의 갯수가 늘어났을 경우 혹은 error처리에 대한 가독성 및 유지 보수를 용이하게 할 수 있는 방법 또한 알고계시면 좋을 것 같습니다. nil 처리, do-cath의 사용, Result Type의 사용이 그 대표적인 예시입니다.

- if 와 guard의 가장 큰 차이는 guard는 반드시 조기 탈출을 하는 곳에 사용해야한다는 점입니다. 이 부분에 대해서만 고려해서 사용하신다면 가독성을 해치지 않는 선에서는 자유롭게 사용하셔도 좋을 것 같습니다.

- 현재 코드에서는 대부분의 예외처리가 guard를 통해서 진행되고 있습니다. 또한 error를 처리하는 반환값 또한 nil로 대부분 같은 편인데요. 비슷한 성질의 error을 한개의 guard로 묶어서 처리하거나 메서드로 추출하여 기능분리를 한다면 지금보다 훨씬 더 가독성이 좋고 이해하기 쉬운 코드가 될 수 있다고 생각합니다.

### 조언을 얻고 싶은 점
1. Error 처리
- 일반적인 Error처리는 Apple에서 소개하고 있는 여러 방법들이 있습니다. 아래 문서의 링크를 첨부합니다.
https://docs.swift.org/swift-book/documentation/the-swift-programming-language/errorhandling/

2. 에러 타입을 커스텀하지 않아도 Error를 전달할 수 있는지
- 질문의 요지를 이해하지 못했습니다. 현재 코드에서는 nil로써 에러를 전달하고 있는 것으로 보입니다. 제가 다시 여쭤보고 싶은 부분은 왜 Error을 전달해야할까요? 더 나아가서는 Error를 어떻게 처리하고 싶으신 건가요?

3. 커밋을 남기는 기준
- 커밋을 남기는 기준은 한 단위의 기능 및 동작 구현, 수정, 리팩토링, 문서 추가가 완료가 되었다면 커밋의 단위로 삼고있습니다만. 매번 커밋 단위를 남기는 것은 습관이 들지 않는다면 생각보다 지켜지기 어렵다고 생각합니다. 저번에 말씀드렸던 GUI Tool을 사용하신다면 문장별로 따로 Stage할 수 있기 때문에 커밋을 나눠서하는데 있어서 훨씬 더 도움이 될 수 있다고 생각합니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
 //
 
+func printMenu() {
+    print("1. 게임 시작")
```
여러개의 print 문의 호출 보다는 하나의 print 함수로도 해결할 수 있는 부분으로 보입니다.

carti1108 replied:
2번의 코멘트와 동일합니다!

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+}
+
+func printGameMenu() {
+    print("숫자 3개를 띄어쓰기로 구분하여 입력해주세요.")
```
위의 코멘트와 같습니다.

Hi-sop replied:
한번에 여러줄을 출력해본 경험이 없어 이렇게 처리했습니다. 
"""를 활용하여 프린트 함수를 한번만 호출하도록 변경했습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+        guard let userInput = readLine() else {
+            continue
+        }
+        switch userInput {
```
저번 프로젝트의 코멘트와 마찬가지로 조건문 앞, 뒤는 개행을 하는 컨벤션으로 보입니다.
해당 부분은 의도하신 부분인가요?

Hi-sop replied:
선언 후 한줄 개행, 함수 마지막 return 전 개행하기로 협의했습니다.
조건문 앞 뒤 개행하는 컨벤션은 따로 생각해보지 않았던거같아요. 저번에도 제가 답변을 애매하게 했네요

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+    
+    while isRunning {
+        printMenu()
+        guard let userInput = readLine() else {
```
이미 userInput이 의도하지 않는 값으로 들어왔을 때에대한 예외처리가 switch에 존재하는 것으로 보입니다. 여기서 OptionalBinding을 사용하신 이유가 있나요?

Hi-sop replied:
코드 작성당시 옵셔널 바인딩이 switch로도 가능한지 인지하지 못했습니다. 수정하도록 하겠습니다!

LeeZion94 replied:
switch는 옵셔널 바인딩을 할 수 없습니다.
다만 Default가 존재할 뿐입니다.

Optional Binding을 할 수 있는 여러가지 방법에 대해 말씀해주실 수 있으실까요?
https://docs.swift.org/swift-book/documentation/the-swift-programming-language/thebasics/#Optional-Binding

Hi-sop replied:
switch는 옵셔널 바인딩이 아니었네요. 이해했습니다!

주로 사용하는 방식은 if-let과 guard-let이 있습니다. 아직 사용해보지 않았지만 while-let도 가능하다고 들었습니다.

LeeZion94 replied:
암시적 추출 방법과 Default 값을 부여하여 Binding을 할 수 있는 방법 또한 있습니다.
확인해보시면 좋을 것 같아요.
https://developer.apple.com/documentation/swift/optional

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+        case "1":
+            playNumberBaseBallGame()
+        case "2":
+            isRunning = false
```
"2"번이 들어왔을 때 break을 주어 반복문을 중단하는 것이 아닌 플래그 값을 사용하신 이유가 있으실까요?

Hi-sop replied:
switch문을 사용할 때 break를 사용하면 switch문만 탈출한다고 생각해 Bool값을 만들어 컨트롤하는 방향으로 작성했습니다.
지금 코드를 다시 보니 case "2": return 으로 변경하면 저희가 의도한 것과 같은 동작을 수행할 수 있다고 생각합니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+    }
+}
+
+func prepareUserInput(userInput: String) -> [Int]? {
```
UserInput을 이미 받은 상황에서 prepare라는 표현은 조금 부정확한 표현으로 보입니다.
다른 표현으로 조금 더 함수의 역활을 명확하게 하는 것은 어떨까요?

Hi-sop replied:
기존에 map을 사용해 unwrap하려다 실패하여 만든 함수였으나(compactMap을 몰랐습니다)
compactMap을 통해 unwrap하는것으로 변경했으므로 해당 함수는 삭제하였습니다.

LeeZion94 replied:
compactMap에 대해 설명해주실 수 있으실까요?

Hi-sop replied:
map과 같이 전달받은 시퀀스를 순회하며 클로저를 적용시킨 결과를 리턴해줍니다. 단 값이 nil인경우는 제외하고 리턴합니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+}
+
+func prepareUserInput(userInput: String) -> [Int]? {
+    let splitUserInput: [Substring] = userInput.split(separator: " ")
```
1. SubString이 무엇이고 어떤 역할을 하고 있나요?
2. Split함수의 기능 및 현재 역할에 대해 설명부탁드립니다.

carti1108 replied:
1. SubString은 String문 안에 일부분을 참조하고 있는것입니다. 참조를 하고있다가 사용되는 시점에 메모리에 올라가는 형태이므로 메모리 관리 측면의 역할을 한다고 생각합니다. 
2. 전달받은 container의 요소를 separator 에 전달한 값을 기준으로 나누고 그 요소들을 배열로 리턴하는 기능입니다. userinput을 띄어쓰기를 기준으로 잘라내어주는 역할을 하고있습니다.

LeeZion94 replied:
훌륭하십니다. 멋지네요.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+    
+    for userNumber in splitUserInput {
+        guard let number = Int(userNumber) else {
+            return nil
```
이미 optionalBinding이 된 Number의 배열이 nil을 반환하기 위해 다시 Optional 배열로 return되고 있습니다.
이 부분을 의도하신 부분이 맞으실까요?

carti1108 replied:
이 부분은 해당 함수였던 prepareUserInput함수를 삭제하고 compactMap함수를 사용하여 수정하였습니다.
수정한 부분은 inputValidate함수로 이동하였습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+}
+
+func inputValidator(userInput: String) -> [Int]? {
+    guard let preparedInput = prepareUserInput(userInput: userInput) else {
```
위의 prepare에 대한 코멘트와 같습니다.
이유가 있는 표현이라면 설명부탁드립니다. 그렇지 않다면 변경해보시는 건 어떨까요?

Hi-sop replied:
입력을 검증한다는 뜻에서 저렇게 지었습니다. 다른 좋은 표현이 있다면 추천해주세요!

LeeZion94 replied:
prepareUserInput의 역할은  UserInput을 전달받아 UserNumbers로 만드는 역할로 보입니다.
위의 내용이 포함된다면 자유롭게 네이밍을 하셔도 될 것 같습니다. 저였다면
convertUserNumbersFromUserInput 으로 해도 적절한 표현으로 보입니다.

다른 이름으로 고민해보시는 건 어떨까요?

Hi-sop replied:
비슷한 역할을 하는 함수를 만들게 되면 추천해주신 네이밍을 참고하여 한번 사용해 보겠습니다. 감사합니다!
현재 prepareUserInput함수는 삭제된 상태입니다

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+    return userNumbers
+}
+
+func inputValidator(userInput: String) -> [Int]? {
```
함수내에 gurad 문이 많아 가독성이 많이 떨어지는 것 같습니다. 반환값은 nil로 통일된다면 비슷한 역할을 하는 guard 끼리는 합치는 것이 어떨까요?

carti1108 replied:
for문을 제외한 guard문 두개는 합쳤습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+    guard preparedInput.count == 3 else {
+        return nil
+    }
+    for value in preparedInput {
```
제가 이해하고 있는 컨벤션과는 다른 표현으로 보입니다. 의도하신 부분이실까요?

carti1108 replied:
위에서 코멘트한 내용과 같은 내용같습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+        return nil
+    }
+    for value in preparedInput {
+        guard preparedInput.filter ({ $0 == value }).count == 1 else {
```
중복된 수를 판단하기 위한 로직으로 보입니다.
for - guard - filter 까지 중첩으로 사용되는 표현으로 가독성이 조금 떨어진다고 생각합니다.
이부분에 있어서는 filter보다는 다른 방식으로 가독성을 높여보는 것이 어떨까요?

Hi-sop replied:
Set으로 타입변경을 하며 중복 숫자를 걸러낼 수 있도록 수정하였습니다.
추가적으로 입력된 수가 3개인지 확인하는 guard와 중복되므로 통합해주었습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+            return nil
+        }
+    }
+    guard preparedInput.filter ({ 1 <= $0 && $0 <= 9 }).count == 3 else {
```
해당 부분은 범위연산자중 ~= 을 사용하신다면 더 쉽게 이해할 수 있는 표현으로 변경될 것 같습니다.

Hi-sop replied:
```swift
1...9 ~= $0
```
형태로 filter의 조건문을 변경했습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
-        print("임의의 수 : \(userRandomNumbers.map { String($0) }.joined(separator: " "))")
-        let (strike, ball) = checkStrikeAndBall(userNumbers: userRandomNumbers, randomNumbers: computerRandomNumbers)
+        printGameMenu()
+        guard let userInput = readLine() else {
```
혹시나 예기치 못하게 else의 부분으로 올경우 유저가 이유를 알 수 있는 방법이 없을 것 같습니다. 이 경우에는 "입력이 잘못되었습니다" 라는 문구 또한 노출되지 않기 때문입니다.
이 경우에 대해서는 optionalBinding을 따로 해주기보다는 inputValidator의 파라미터를 optional로 변경한뒤 inputValidator에서 optional Binding에 대한 처리를 같이 해주는 것이 더 좋은 방법이라고 생각됩니다.
이 부분에 대해서 어떻게 생각하시나요?

Hi-sop replied:
말씀하신 내용이 더 합리적으로 느껴져 변경하기로 협의했습니다.
기존 playNumberBaseBallGame 에서 언랩하던 부분을 inputValidator 처리하도록 변경했습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+    return preparedInput
+}
 
 func playNumberBaseBallGame() {
```
함수의 길이가 약 30줄로 생각보다 긴 것 같습니다.
함수의 길이가 길어질 때는 하나의 함수에서 여러가지 일을 하는 것은 아닌가? 라는 생각을 해보시는 것도 좋을 것 같습니다,
User의 입력을 받는 부분, User 입력으로 Strike, Ball을 체크하는 부분 크게 두가지로 나눠질 수 있다고 판단됩니다.
혹시 나눠보시는 것은 어떨까요?

carti1108 replied:
playNumberBaseBallGame함수에서 게임결과를 판단하는 로직을 다른 함수로 분리해주었습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
 //
 //  NumberBaseball - main.swift
-//  Created by yagom.
+//  Created by Kiseok, Hisop.
```
main 파일은 clone시 존재하고 있는 파일로 알고 있습니다.
생성자를 변경하신 이유가 있으실까요?

carti1108 replied:
사실 아무 생각없이 수정하였습니다,, 수정하면 안되는 것이라면 다음부터는 수정 안하겠습니다!

LeeZion94 replied:
이미 존재하는 파일에 대해서는 히솝, 기석께서 생성자가 아니시기 때문에 변경하실 필요는 없지만, 프로젝트를 하시면서
새롭게 만드시는 파일에 대해서는 생성자를 히솝, 기석으로 수정해주시면 될 것 같습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,24 +1,92 @@
+    var userNumbers: [Int] = []
+    
+    for userNumber in splitUserInput {
+        guard let number = Int(userNumber) else {
```
1. Optional이란 무엇인가요? 또한 어떠한 것들로 이루어져있나요?
2. Optional은 왜 필요할까요?

carti1108 replied:
1. Optional은 enum타입으로 이루어져 있고 값이 있을수도 있고 없을수도 있다는 것을 나타내는 타입입니다.
2. 스위프트의 특성 중 하나는 안전성 이기 때문에 런타임 오류를 내지 않도록 값이 없음을 나타내는 nil 타입을 정의하기 위해 필요하다고 생각합니다.

LeeZion94 replied:
1. Optional은 enum으로만 이뤄진 것이 아닌 enum과 Generic이 합쳐져 있는 구조로 이루어져 있습니다.
- 그렇다면 Generic이란 무엇이고, 왜 필요한가요?

carti1108 replied:
Generic은 어떤 타입이든 받을 수 있다고 선언을 해줄때 사용됩니다. Generic을 사용하게 되면 Generic으로 구현된 기능을 쉽게 재사용 할 수 있고, 코드의 중복도 줄여줄 수 있어 깔끔하고 유연한 코드의 작성이 가능해지도록 하기 때문에 필요합니다.

carti1108 committed:
refactor: 네이밍 수정 및 일부 함수 삭제

print함수를 한번만 호출하도록 변경
중복된 optional unwraping 삭제
반복문 탈출을 flag값이 아닌 return을 사용하도록 변경
prepareUserInput함수를 삭제하고 compactMap함수를 사용하도록 변경

Commit:
style: guard문 통합

inputValidate함수 내부의 guard문 통합

carti1108 committed:
refactor: 중복된 수를 판단하는 로직 변경, 사용자 입력 optional 처리 변경

중복된 수를 판단하는 로직을 filter함수에서 set으로 변경
set으로 변경하며 중복되는 guard문을 통합
filter의 범위 연산자 변
playNumberBaseBall함수에서 처리하던 사용자 입력을 inputValidate함수에서 처리하도록 변경

Commit:
refactor: 게임결과 판단 로직 함수로 분리

playNumberBaseBallGame 함수에서 게임 결과 판단 로직을 gameResult 함수로 분리

carti1108 committed:
style: 코드 컨벤션 수정(개행)

LeeZion94 commented:
고생 정말 많이 하셨습니다.
